File: Julia/GeneralFunctions/normalMLE.jl
@views @inline function normalMLE(bstart::Vector,Y::Array,X::Array,W::Array=ones(length(Y)),d::Array=ones(length(Y)))

    J = length(unique(d))
    n = length(Y)
 
    # error checking
    @assert size(X,1)==n "The 1st dimension of X should equal the number of observations in Y"
    @assert minimum(d)==1 && maximum(d)==J  "d should contain integers numbered consecutively from 1 through J"
    @assert size(X,2)+J==size(bstart,1) "parameter vector has wrong number of elements"
    @assert size(W,1)==n "The size of the weighting vector should equal the number of observations in Y"
 
    function f(b)
        T         = promote_type(promote_type(eltype(X),eltype(b)),eltype(W))
        dmat      =  ones(T,n,J)
        likemat   = zeros(T,n,J)
        liker     = zeros(T,n)
        ℓ         =  zero(T)
        beta      = b[1:end-J]
        wagesigma = b[end-(J-1):end]
        for j=1:J
            dmat[:,j] .= d.==j
            likemat[:,j] .= vec( -0.5.*( log.(2 .* pi) .+ log.(wagesigma[j]^2) .+ ( (Y.-X*beta)./wagesigma[j] ).^2 ) )
        end
        liker .= vec(sum(dmat.*likemat;dims=2))
        ℓ = -W'*liker
    end

    function g!(G,b)
        T         = promote_type(promote_type(promote_type(eltype(X),eltype(b)),eltype(Y)),eltype(W))
        temp      = zeros(T,size(Y,1))
        beta      = b[1:end-J]
        wagesigma = b[end-(J-1):end]

        G .= T(0)
        for j=1:J
            G[1:end-J] .+= vec( -X'*( W.*(d.==j).*(Y.-X*beta)./(wagesigma[j].^2) ) )
        end
        for j=1:J
            k=length(b)-(J-1)+j-1
            temp .= vec( 1 ./ wagesigma[j] .- ( (Y.-X*beta).^2) ./ (wagesigma[j].^3 ) )
            G[k] = sum(W.*(d.==j).*temp)
        end
        return nothing
    end

    td = TwiceDifferentiable(f, g!, bstart, autodiff = :forwarddiff)
    rs = optimize(td, bstart, LBFGS(; linesearch = LineSearches.BackTracking()), Optim.Options(iterations=100_000,g_tol=1e-6,f_tol=1e-6))
    β  = Optim.minimizer(rs)
    ℓ  = Optim.minimum(rs)*(-1)
    H  = Optim.hessian!(td, β)
    g  = Optim.gradient!(td, β)
    se = sqrt.(diag(inv(H)))

    return β,se,ℓ,g
end

File: Matlab/GeneralFunctions/normalMLE.m
function [like,grad]=normalMLE(b,restrMat,Y,X,W,d)
%NORMALMLE estimates a linear regression model with potentially hetero-
%   skedastic error variances.
%   
%   LIKE = NORMALMLE(B,RESTRMAT,Y,X,D) 
%   estimates a linear regression model with errors assumed to be normal. 
%   The user can specify heteroskedasticity in these error variances.
%   Parameter restrictions are constructed in RESTRMAT.
%   
%   For estimation without restrictions, set RESTRMAT to be an empty matrix. 
%   
%   RESTRMAT is an R x 5 matrix of parameter restrictions. See APPLYRESTR
%   for more information in using this feature. If no parameter
%   restrictions are desired, RESTRMAT should be passed as an empty matrix
%   Y is an N x 1 vector of outcomes. 
%   X is an N x K matrix of covariates.
%   W is an N x 1 vector of weights.
%   D is an N x 1 vector of integers that indicates which variance group
%   an observation falls into (heteroskedasticity case). If homoskedastic
%   errors are assumed, D may be left unpassed or passed as an empty matrix
%   B is the parameter vector, with K + numel(unique(D)) elements
%   
%   This function does *not* automatically include a column of ones in X.
%   It also does *not* automatically drop NaNs
 
% Copyright 2014 Tyler Ransom, Duke University
% Revision History: 
%   September 25, 2014
%     Created 
%   September 29, 2014
%     Added code for weighted estimation 
%==========================================================================
 
% error checking
assert(size(X,1)==size(Y,1),'X and Y must be the same length');
if nargin==5
    d = ones(size(Y));
elseif nargin==6 && isempty(d)
    d = ones(size(Y));
end
J = numel(unique(d));
assert(  min(d)==1 && max(d)==J   ,'d should contain integers numbered consecutively from 1 through J');
assert(size(X,2)+J==size(b,1),'parameter vector has wrong number of elements');
if ~isempty(W)
    assert(isvector(W) && length(W)==length(Y),'W must be a column vector the same size as Y');
else
    W = ones(size(Y));
end
 
% apply restrictions as defined in restrMat
if ~isempty(restrMat)
    b = applyRestr(restrMat,b);
end
 
% slice parameter vector
beta      = b(1:end-J);
wagesigma = b(end-(J-1):end);
n         = length(Y);
 
% log likelihood
likemat = zeros(n,J);
dmat    = zeros(n,J);
for j=1:J
    dmat(:,j) = d==j;
    likemat(:,j) = -.5*(log(2*pi)+log(wagesigma(j)^2)+((Y-X*beta)./wagesigma(j)).^2);
end
like = -W'*sum(dmat.*likemat,2);
 
% analytical gradient
grad = zeros(size(b));
for j=1:J
    grad(1:end-J) = -X'*(W.*(d==j).*(Y-X*beta)./(wagesigma(j).^2)) + grad(1:end-J);
end
for j=1:J
    k=length(b)-(J-1)+j-1;
    temp = 1./wagesigma(j)-((Y-X*beta).^2)./(wagesigma(j).^3);
    grad(k) = sum(W.*(d==j).*temp);
end
 
end